File: packages/twerk-slots/src/twerk_slots/cli/slot/goto.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated

import click

from twerk_core import get_console
from twerk_core.clinkr.dataclass_json import JsonSerializable
from twerk_core.clinkr.exit import ClinkrExit
from twerk_core.clinkr.operation import clinkr_operation
from twerk_slots.cli.slot.context import load_slots_context
from twerk_slots.cli.slot.selectors import SelectorOk, resolve_num, resolve_wt
from twerk_slots.inventory import build_slot_inventory
from twerk_slots.repo_context import NoRepoSentinel


@dataclass(frozen=True)
class SlotGotoRequest:
    num: Annotated[int | None, click.Option(["--num"], type=click.INT, default=None)] = None
    wt: Annotated[str | None, click.Option(["--wt"], type=click.STRING, default=None)] = None


@dataclass(frozen=True)
class SlotGotoResult(JsonSerializable):
    slot_name: str
    branch_name: str
    worktree_path: str


def render_slot_goto(result: SlotGotoResult) -> None:
    console = get_console()
    console.print(
        f"[bold cyan]{result.slot_name}[/bold cyan] -> [green]{result.branch_name}[/green]"
    )
    # Pipeable worktree path on its own line (last).
    click.echo(result.worktree_path)


@clinkr_operation(
    name="goto",
    help="Print the worktree path for an assigned slot.",
    human_renderer=render_slot_goto,
)
def run_goto_slot(ctx: click.Context, request: SlotGotoRequest) -> ClinkrExit[SlotGotoResult]:
    slots_ctx = load_slots_context(ctx)
    if isinstance(slots_ctx, NoRepoSentinel):
        raise ClinkrExit.failure(error_type="not_in_repo", message=slots_ctx.message)

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    if inventory.pool_size == 0:
        raise ClinkrExit.failure(
            error_type="pool_empty",
            message="No managed slots configured. Run `slot init --size N` first.",
        )

    if request.num is not None and request.wt is not None:
        raise ClinkrExit.failure(
            error_type="conflicting_slot_args",
            message="Pass exactly one of --num or --wt, not both.",
        )
    if request.num is not None:
        result = resolve_num(request.num, inventory.pool_size)
        if not isinstance(result, SelectorOk):
            raise ClinkrExit.failure(error_type="invalid_slot_num", message=result.message)
        slot_name = result.slot_name
    elif request.wt is not None:
        result = resolve_wt(request.wt)
        if not isinstance(result, SelectorOk):
            raise ClinkrExit.failure(error_type="invalid_slot_wt", message=result.message)
        slot_name = result.slot_name
    else:
        raise ClinkrExit.failure(
            error_type="missing_slot_arg",
            message="Pass one of --num or --wt to identify the slot.",
        )

    record = inventory.find_by_slot(slot_name)
    if record is None or record.branch is None:
        raise ClinkrExit.negative(
            message=f"{slot_name} is not currently assigned. Run `slot list` to see the pool.",
        )

    if not slots_ctx.storage.path_exists(record.path):
        raise ClinkrExit.failure(
            error_type="worktree_missing",
            message=(
                f"Worktree for {slot_name} is missing at {record.path}. "
                f"Run `slot free --wt {slot_name}` to clear the stale assignment."
            ),
        )

    return ClinkrExit.ok(
        SlotGotoResult(
            slot_name=slot_name,
            branch_name=record.branch,
            worktree_path=str(record.path),
        )
    )

File: ts/packages/capabilities/slot/src/operations/goto.ts
import { failure, negative, ok, type RenderCapabilities } from "@sdl/clinkr";
import { z } from "zod";

import type { RepoSlotContext, SlotCliContext } from "../core/context.ts";
import { buildSlotInventory, findBySlot } from "../core/inventory.ts";
import { prepareNavigation } from "../core/navigation-result.ts";
import { renderSlotNavigationSuccess } from "../core/navigation-presentation.ts";
import { poolSize } from "../core/inventory.ts";
import { resolveNum, resolveWt } from "../core/selectors.ts";

export const gotoRequestSchema = z.object({
	num: z.number().int().optional().describe("Slot number."),
	wt: z.string().optional().describe("Slot worktree name, e.g. slot-01."),
	clipboard: z.boolean().default(true).describe("Copy the cd command to the clipboard."),
});

export const gotoResultSchema = z.object({
	slotName: z.string(),
	branchName: z.string(),
	operation: z.string().nullable(),
	worktreePath: z.string(),
	cdCommand: z.string(),
	clipboardCopied: z.boolean(),
	clipboardSkipped: z.boolean(),
	clipboardFailureReason: z
		.union([z.literal("backend-missing"), z.literal("subprocess-error")])
		.nullable(),
	clipboardFailureDetail: z.string().nullable(),
});

export type GotoRequest = z.infer<typeof gotoRequestSchema>;
export type GotoResult = z.infer<typeof gotoResultSchema>;

export async function runGoto(ctx: SlotCliContext, request: GotoRequest) {
	if (ctx.repo.type !== "repo") return failure(ctx.repo.errorType, ctx.repo.message);
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	const inventory = await buildSlotInventory(repoCtx.git, {
		mainRepoRoot: repoCtx.repo.mainRepoRoot,
	});
	if (poolSize(inventory) === 0)
		return failure("pool-empty", "No managed slots configured. Run `slot init --size N` first.");
	if (request.num !== undefined && request.wt !== undefined)
		return failure("conflicting-slot-args", "Pass exactly one of -n/--num or -w/--wt, not both.");
	let slotName: string;
	if (request.num !== undefined) {
		const result = resolveNum(request.num, poolSize(inventory));
		if (result.type === "error") return failure("invalid-slot-num", result.message);
		slotName = result.slotName;
	} else if (request.wt !== undefined) {
		const result = resolveWt(request.wt);
		if (result.type === "error") return failure("invalid-slot-wt", result.message);
		slotName = result.slotName;
	} else {
		return failure("missing-slot-arg", "Pass one of -n/--num or -w/--wt to identify the slot.");
	}
	const record = findBySlot(inventory, slotName);
	if (record === null || record.branch === null)
		return negative(`${slotName} is not currently assigned. Run \`slot list\` to see the pool.`);
	if (!(await repoCtx.git.pathExists(record.path)))
		return failure(
			"worktree-missing",
			`Worktree for ${slotName} is missing at ${record.path}. Run \`sdl slot free --wt ${slotName}\` to clear the stale assignment.`,
		);
	const navigation = await prepareNavigation(repoCtx, record.path, {
		shouldCopyClipboard: request.clipboard,
		shouldWriteCdDirective: repoCtx.shouldWriteCdDirective,
	});
	return ok({
		slotName: slotName,
		branchName: record.branch,
		operation: record.operation,
		...navigation,
	});
}

export function renderGoto(
	result: GotoResult,
	caps: RenderCapabilities = { canEmitAnsi: false },
): string {
	const operationSuffix = result.operation === null ? "" : ` (${result.operation} in progress)`;
	return renderSlotNavigationSuccess(
		{
			...result,
			headline: `${result.slotName} -> ${result.branchName}${operationSuffix}`,
		},
		caps,
	);
}
